daisy/step_copy_gcs_objects.go

```diff
@@ -31,14 +31,88 @@ type CopyGCSObjects []CopyGCSObject
 // CopyGCSFile copies a GCS file from Source to Destination.
 type CopyGCSObject struct {
 	Source, Destination string
-	ACLRules            []storage.ACLRule
+	ACLRules            []*storage.ACLRule
 }
 
-func (c *CopyGCSObjects) populate(ctx context.Context, s *Step) error { return nil }
+func (c *CopyGCSObjects) populate(ctx context.Context, s *Step) error {
+	for _, co := range *c {
+		for _, acl := range co.ACLRules {
+			acl.Role = storage.ACLRole(strings.ToUpper(string(acl.Role)))
+		}
+	}
+	return nil
+}
 
-func (c *CopyGCSObjects) validate(ctx context.Context, s *Step) error { return nil }
+func (c *CopyGCSObjects) validate(ctx context.Context, s *Step) error {
+	for _, co := range *c {
+		sBkt, _, err := splitGCSPath(co.Source)
+		if err != nil {
+			return err
+		}
+		dBkt, _, err := splitGCSPath(co.Destination)
+		if err != nil {
+			return err
+		}
+
+		// Check if source bucket exists and is readable.
+		if !readableBkts.contains(sBkt) {
+			if _, err := s.w.StorageClient.Bucket(sBkt).Attrs(ctx); err != nil {
+				return fmt.Errorf("error reading bucket %q: %v", sBkt, err)
+			}
+			readableBkts.add(sBkt)
+		}
 
-func recursiveGCS(ctx context.Context, w *Workflow, sBkt, sPrefix, dBkt, dPrefix string) error {
+		// Check if destination bucket exists and is readable.
```

Comment:
Why does destination bucket need to be readable? Why not just a writable check?

Reply:
The errors you get from the read check are much more readable, there's a nice difference between 'access denied' and 'bucket does not exist' and with an access denied you get a lot of useful info about the user